demoapp/demo_app_functions.py

```diff
@@ -450,4 +512,642 @@ def update_csv(bill_num: str, title: str, summarized_bill: str, category: str, t
+
+    Arguments: 
+
+    bill_details (BillDetails): Object containing information about the bill - bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+
+    Returns: 
+        A tuple of status_code and an LLMResults object containing query, response from the LLM
+        status_code can take these following values {1: Success, -1: Necessary details not found}
+
+    '''
+
+    if not all(hasattr(bill_details, attr) for attr in ("bill_text", 'bill_title', 'mgl_names', 'committee_info')): 
+        return -1, LLMResults()
+
+    set_my_llm_cache()
+    llm_call_type = get_llm_call_type(bill_details)
+
+    query_1 = get_query_for_categorizing(bill_details, llm_call_type)
+    category_response = call_llm(bill_details, query_1, llm_call_type)
+    categories = extract_categories_tags(category_response.response)
+    category_tags = get_category_tags(categories)
+    query_2 = get_query_for_tagging(bill_details, category_tags, llm_call_type)
+    tag_response = call_llm(bill_details, query_2, llm_call_type)
+
+    # parses the response from LLM and removes hallucinated tags
+    tag_response.response = list(set(extract_categories_tags(tag_response.response)) & set(category_tags))
+
+    return 1, tag_response
+
+def extract_categories_tags(response: str):
+    response = response.split('#')
+    return [i.strip() for i in response]
+
+def prepare_invoke_dict(bill_details: BillDetails) -> dict:
+
+    """
+    This function prepares the dict object that is used in chain.invoke function to call the LLM with prompt and 
+    required details. 
+
+    Args: 
+        bill_details (BillDetails): Object containing information about the bill - bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+
+    Returns: 
+        dict object containing all the necessary keys and values required for invoke call. 
+
+    """
+    text_splitter = CharacterTextSplitter(chunk_size = 90000, chunk_overlap = 1000)
+
+    return {
+                "title": bill_details.bill_title, 
+                "context": [Document(page_content = f"```{x}```") for x in text_splitter.split_text(bill_details.bill_text)],
+                "names": bill_details.mgl_names, 
+                "mgl_sections": [Document(page_content = f"```{x}```") for x in text_splitter.split_text(bill_details.mgl_ref)],
+                "committee_info": bill_details.committee_info
+            }
+
+def get_query_for_summarization(bill_details: BillDetails, llm_call_type: str) -> str:
+
+    """
+        
+    This functions prepares a prompt based on the call type (small: No use of RAG, large: Use RAG) for bill summarization
+
+    Args: 
+        bill_details (BillDetails): Object containing information about the bill - bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+        llm_call_type (str): This argument can take 2 values ("small": No use of RAG, "large": Use RAG) 
+
+    Returns: 
+        str: Query that includes task, context and instructions on summary generation
```

Comment:
Does this also modify `bill_details` inplace? If so, document here.

Reply:
Yes, it does. I will document it in the doc string